homeassistant/components/sensor/bom.py

```diff
@@ -66,25 +74,53 @@
     'wind_spd_kt': ['Wind Direction kt', 'kt']
 }
 
+
+def validate_station(station):
+    """Check that the station ID is well-formed."""
+    if station is None:
+        return
+    station = station.replace('.shtml', '')
+    if not re.fullmatch(r'ID[DNQTVW]60801\.\d\d\d\d\d', station):
+        raise vol.error.Invalid('Malformed station ID')
+    return station
+
+
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
-    vol.Required(CONF_ZONE_ID): cv.string,
-    vol.Required(CONF_WMO_ID): cv.string,
+    vol.Optional(CONF_ZONE_ID): cv.string,
+    vol.Optional(CONF_WMO_ID): cv.string,
     vol.Optional(CONF_NAME, default=None): cv.string,
+    vol.Optional(CONF_STATION): validate_station,
     vol.Required(CONF_MONITORED_CONDITIONS, default=[]):
         vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
 })
 
 
 def setup_platform(hass, config, add_devices, discovery_info=None):
     """Set up the BOM sensor."""
-    rest = BOMCurrentData(
-        hass, config.get(CONF_ZONE_ID), config.get(CONF_WMO_ID))
-
-    sensors = []
-    for variable in config[CONF_MONITORED_CONDITIONS]:
-        sensors.append(BOMCurrentSensor(
-            rest, variable, config.get(CONF_NAME)))
+    station = config.get(CONF_STATION)
+    zone_id, wmo_id = config.get(CONF_ZONE_ID), config.get(CONF_WMO_ID)
+    if station is not None:
+        pass
+    elif zone_id and wmo_id:
+        station = zone_id + '.' + wmo_id
+    elif wmo_id:
+        stations = bom_stations(hass.config.config_dir)
+        lookup = {wmo: zone for zone, wmo in (k.split('.') for k in stations)}
+        station = lookup[wmo_id] + '.' + wmo_id
+    else:
+        if zone_id:
+            _LOGGER.info('zone_id does not uniquely identify a station, '
+                         'falling back to lat/lon autodetection...')
+        station = closest_station(config.get(CONF_LATITUDE),
+                                  config.get(CONF_LONGITUDE),
+                                  hass.config.config_dir)
+        if station is None:
+            _LOGGER.error("Could not get BOM weather station from lat/lon")
+            return False
 
+    rest = BOMCurrentData(hass, station)
+    sensors = [BOMCurrentSensor(rest, variable, config.get(CONF_NAME))
+               for variable in config[CONF_MONITORED_CONDITIONS]]
     try:
```

Comment:
I'm guessing...wouldn't it be more efficient to check the data object before we are creating the sensors?